Allow a grid with only 1 row

mlms13 opened the pull request:
Don't merge this yet, but I'm working on a fix for #5.

Part of the problem was that we were assuming that `fixedTop + 1` and `fixedBottom - 1` were different rows. The easiest way to avoid this problem is to just iterate over all of the rows and filter out the ones we don't care about. Are we worried about the speed implications of this complete `for` loop with in `if` at the top of it?

mlms13 committed:
Allow a grid with only 1 row

mlms13 committed:
Add more refreshing to the setRowsAndCols method

This might be overkill, and even if it isn't, there are a lot of things here that can be refactored. Plus, a big ol' FIXME comment in Grid9, where I'm clearly doing things wrong. But hey, at least it allows FancyTable to work.

fponticelli commented on src/fancy/Grid.hx:
```diff
@@ -147,10 +147,23 @@ class Grid {
     this.rows = rows;
```
we could probably call `setRowsAndColumns` from the constructor instead of repeating everything twice.

mlms13 replied:
You didn't like my disgusting copy-and-paste?

fponticelli commented on src/fancy/Grid.hx:
```diff
@@ -147,10 +147,23 @@ class Grid {
     renderMiddle(grid9.position.y);
     renderCenter(grid9.position.x);
     renderMain(grid9.position.x, grid9.position.y);
+    grid9.refresh();
```
`refresh()` is also called from the `Grid9` constructor ... so if really needed here, maybe we remove that call from the constructor?

fponticelli commented on src/fancy/Grid.hx:
```diff
@@ -190,32 +203,35 @@ class Grid {
+      for (i in 0...rows) {
+        // render each row that is in the fixed top, or the first row after
+        // fixed top, or the last row before fixed bottom, or in fixed bottom
+        if (i < fixedTop + 1 || i > rows - fixedBottom - 1) {
```
you killed my optimization ... with 1.000.000 rows this check will be performed that many times. It is certainly a fast check so maybe it is fine this way.

mlms13 replied:
I know I did, and we should probably switch back, but it was wrong previously for 2x2 grids (with either a fixed header or footer). If the first row of content was the same as the last row of content, it would try to measure it and remove it from the parent twice. The second time, it would fail, because it had already removed that element from the parent.

This was my quick and dirty solution to at least get it to do the right thing. It's definitely not as efficient, so we should change back to something closer to what you had before.

fponticelli replied:
I see your point. We can make it work :)

fponticelli commented on src/fancy/Grid.hx:
```diff
@@ -190,32 +203,35 @@ class Grid {
+      // then remove all of the elements (separate from measuring, for speed)
       for(el in els)
         view.removeChild(el);
+
```
I don't approve the empty line ... it is sooo empty.

fponticelli commented on src/fancy/core/Grid9.hx:
```diff
@@ -312,8 +312,11 @@ class Grid9 {
   public function sizeRails(topRail: Float, bottomRail: Float, leftRail: Float, rightRail: Float) {
-    if(this.topRail == topRail && this.bottomRail == bottomRail && this.leftRail == leftRail && this.rightRail == rightRail)
-      return;
+    // FIXME: i commented out the early return because sometimes this function
```
I don't understand what the problem is but this might need to be restored at some point.

mlms13 replied:
This is the only function that updates the `contentHeight` and `contentWidth` styles in the DOM, but the guard was causing it to return early if rails hadn't changed. But in a case where rails hadn't changed but contentheight/width had, i still want to hit all of the DOM updates below. Really, we should probably split this whole function into two.

fponticelli replied:
Agreed ... the split needs to happen. Or move the resizing in refresh ... who know really?

mlms13 commented on src/fancy/Grid.hx:
```diff
@@ -147,10 +147,23 @@ class Grid {
+    bottomRailSize = fixedBottom == 0 ? 0 : (contentHeight - vOffset(rows - fixedBottom));
+    rightRailSize = fixedRight == 0 ? 0 : (contentWidth - hOffset(columns - fixedRight));
+    grid9.resizeContent(contentWidth, contentHeight);
+    grid9.sizeRails(topRailSize, bottomRailSize, leftRailSize, rightRailSize);
```
@fponticelli: rails haven't actually changed here. We're just calling this function for the side effect of updating height/width dimensions in the element styles. This is why I removed the early return in `sizeRails`, but obviously there should be a better way.

mlms13 committed:
Add several size-calculating strategies

mlms13 committed:
Fix binary search for out-of-bound values

mlms13 committed:
Scroll correctly after sizing down content

mlms13 committed:
Always set horizontal size before measuring vertically

mlms13 committed:
Allow touch events to turn into mouse events

If we prevent default on touchstart/end, it stops the browser from firing mouse events (like click) in response touch actions.

mlms13 committed:
Remove duplication, and clean up grid9 resize methods

mlms13 committed:
Allow setting scroll position from the outside

mlms13 commented:
Apparently we've been using the latest commit here (`19ff94a`) in production for 7 months now. There are definitely ways to make the code a little cleaner, but I'm going to merge what's here.
